src/t_hash.c

```diff
@@ -2030,3 +2078,124 @@ void hrandfieldCommand(client *c) {
     hashTypeRandomElement(hash, hashTypeLength(hash), &ele, NULL);
     hashReplyFromListpackEntry(c, &ele);
 }
+
+/* Context structure for tracking expiry operations on hash fields. */
+typedef struct {
+    robj *key;               /* the hash object */
+    unsigned long n_entries; /* number of entries processed */
+    void **entries;          /* array of expired entries to replicate and free later */
+} expiryContext;
+
+/* Callback for popping expired entries from the volatile set.
+ * Deletes the entry from the hash table and tracks it in the expiry context.
+ * Returns 1 if deleted, 0 if nothing to do. */
+static int expireEntry(void *entry, void *c) {
+    expiryContext *ctx = c;
+    robj *o = ctx->key;
+    serverAssert(o->encoding == OBJ_ENCODING_HASHTABLE);
+
+    hashtable *ht = o->ptr;
+    void *entry_ptr = NULL;
+    int deleted = hashtablePop(ht, entry, &entry_ptr);
+
+    if (deleted) {
+        ctx->entries[ctx->n_entries++] = entry_ptr;
+        server.stat_expiredfields++;
+        return 1;
+    }
+    return 0;
+}
+
+/* Extract expired entries from a hash object's volatile set.
+ * Returns number of expired entries, populates `out_entries`. */
+size_t hashTypePopExpiredFields(robj *o, mstime_t now, unsigned long max_entries, void **out_entries) {
+    serverAssert(o->encoding == OBJ_ENCODING_HASHTABLE);
+    serverAssert(max_entries > 0 && max_entries <= 1024);
+
+    /* skip TTL checks temporarily (to allow hashtable lookup) */
+    hashTypeIgnoreTTL(o, 1);
+
+    vset *vset = hashTypeGetVolatileSet(o);
+    if (vsetIsEmpty(vset)) {
+        hashTypeIgnoreTTL(o, 0);
+        return 0;
+    }
+
+    expiryContext ctx = {.key = o, .entries = out_entries, .n_entries = 0};
+    size_t expired = vsetPopExpired(vset, entryGetExpiry, expireEntry, now, max_entries, &ctx);
+    serverAssert(ctx.n_entries <= max_entries);
+    hashTypeIgnoreTTL(o, 0);
+    if (!hashTypeHasVolatileFields(o)) {
+        hashTypeFreeVolatileSet(o);
+    }
+    return expired;
+}
+
+/* Hashtable scan callback for hash datatype */
+static void defragHashTypeEntry(void *privdata, void *element_ref) {
+    entry **entry_ref = (entry **)element_ref;
+    entry *old_entry = *entry_ref, *new_entry = NULL;
+    long long old_expiry = entryGetExpiry(old_entry);
+
+    new_entry = entryDefrag(*entry_ref, activeDefragAlloc, activeDefragSds);
```

Comment:
This threw me off for a sec.  Why are you using `entry_ref` when you've already dereferenced it into `old_entry`?
```suggestion
    entry *new_entry = entryDefrag(old_entry, activeDefragAlloc, activeDefragSds);
```